### DISCUSSION TITLE
IconSource transforms

### DISCUSSION THREAD

**Discussion by stevemonaco** [amwx/FluentAvalonia#327] [Q&A] [answered]
I'm upgrading an app from 2-preview4 to 2-preview6 and I'm having some difficulties with the new `IconSource`. In particular, when parent controls have an `IconSource` dependency property, I cannot figure out how to apply a transform to only the icon. Transforms are important to easily lift SVG paths from the MahApps.IconPacks browser because many are stored mirrored, flipped, or turned (both).

There's probably a way to apply a transform out-of-app (Inkscape?), but I can't figure out a reasonable workflow. If there's no workaround and external app support is weak here, maybe a feature request is warranted.

Old (2-preview4):
```xml
<DataTemplate x:Key="{x:Type vm:HistoryViewModel}" DataType="vm:HistoryViewModel">
	<ui:NavigationViewItem Content="{Binding DisplayName}">
		<ui:NavigationViewItem.Icon>
			<ui:FAPathIcon Data="{StaticResource iconHistory}" RenderTransform="scaley(-1)"/>
		</ui:NavigationViewItem.Icon>
	</ui:NavigationViewItem>
</DataTemplate>
```

New (2-preview6, no icon rotation):
```xml
<DataTemplate x:Key="{x:Type vm:HistoryViewModel}" DataType="vm:HistoryViewModel">
	<ui:NavigationViewItem Content="{Binding DisplayName}">
		<ui:NavigationViewItem.IconSource>
			<ui:PathIconSource Data="{StaticResource iconHistory}" />
		</ui:NavigationViewItem.IconSource>
	</ui:NavigationViewItem>
</DataTemplate>
```

![image](https://user-images.githubusercontent.com/1782158/228057974-8bef5a96-0f68-4951-bcb9-e22e514630a2.png)

Where `iconHistory` looks like:
```xml
<StreamGeometry x:Key="iconHistory">M854.173 <Omitted...> 3.182-0.748-1.084-1.533-2.138-2.327-3.182z</StreamGeometry
```

If I apply a `RenderTransform="scaley(-1)"` to `NavigationViewItem` to correctly orientate the scroll icon, the header content is flipped as well.

![image](https://user-images.githubusercontent.com/1782158/228054880-f6311574-ffad-4ec8-9e04-627e8dab9177.png)

**Comment by amwx** [comment, marked as answer]
As a workaround for now, you should be able to use a Style Selector. A normal FAPathIcon is still generated from the IconSource:
```xaml
<Style Selector="ui|NavigationViewItem ui|FAPathIcon">
    <Setter Property="RenderTransform" ... />
</Style>
```
(note: I haven't tested this, but this should work. If that doesn't work, you could try `uip|NavigationViewItemPresenter ui|FAPathIcon`, where uip is the primitives namespace)

Might be possible to add a Transform property to IconSource or maybe just PathIconSource to handle this. Need to think on this (more the implementation details, this is a strong argument to include it)...not sure if I'll get it into v2 though.

**Comment by stevemonaco** [comment]
Ah, nice workaround. It does work and can be applied per-item if needed. Thanks!

```xml
<DataTemplate x:Key="{x:Type vm:HistoryViewModel}" DataType="vm:HistoryViewModel">
	<ui:NavigationViewItem Content="{Binding DisplayName}">
		<ui:NavigationViewItem.Styles>
			<Style Selector="ui|NavigationViewItem ui|FAPathIcon">
				<Setter Property="RenderTransform" Value="scaley(-1)" />
			</Style>
		</ui:NavigationViewItem.Styles>
		<ui:NavigationViewItem.IconSource>
			<ui:PathIconSource Data="{StaticResource iconHistory}" />
		</ui:NavigationViewItem.IconSource>
	</ui:NavigationViewItem>
</DataTemplate>
```

Or a bit more tidy:
```xml
<Style Selector="ui|NavigationViewItem.flip ui|FAPathIcon">
	<Setter Property="RenderTransform" Value="scaley(-1)" />
</Style>
...
<DataTemplate x:Key="{x:Type vm:HistoryViewModel}" DataType="vm:HistoryViewModel">
	<ui:NavigationViewItem Classes="flip" Content="{Binding DisplayName}">
		<ui:NavigationViewItem.IconSource>
			<ui:PathIconSource Data="{StaticResource iconHistory}" />
		</ui:NavigationViewItem.IconSource>
	</ui:NavigationViewItem>
</DataTemplate>
```
